writer_test.go

```diff
@@ -90,3 +92,63 @@ func TestWritePolyLine(t *testing.T) {
 	}
 	testPolyLine(t, pointsToFloats(flatten(points)), shapes)
 }
+
+type seekCloseTracker struct {
+	io.Writer
+	offset int64
+	closed bool
+}
+
+func (s *seekCloseTracker) Seek(offset int64, whence int) (int64, error) {
+	s.offset = offset
+	return 0, nil
+}
+
+func (s *seekCloseTracker) Close() error {
+	s.closed = true
+	return nil
+}
+
+func TestWriteAttribute(t *testing.T) {
+	buf := new(bytes.Buffer)
+	s := &seekCloseTracker{Writer: buf}
+	w := Writer{
+		dbf: s,
+		dbfFields: []Field{
+			StringField("A_STRING", 6),
+			FloatField("A_FLOAT", 8, 4),
+			NumberField("AN_INT", 4),
+		},
+		dbfRecordLength: 100,
+	}
+
+	tests := []struct {
+		name       string
+		row        int
+		field      int
+		data       interface{}
+		wantOffset int64
+		wantData   string
+	}{
+		{"string-0", 0, 0, "test", 1, "test"},
+		{"string-3", 3, 0, "things", 301, "things"},
+		{"float-0", 0, 1, 123.44, 7, "123.4400"},
+		{"int-0", 0, 2, 4242, 15, "4242"},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			buf.Reset()
+			s.offset = 0
+
+			w.WriteAttribute(test.row, test.field, test.data)
+
+			if buf.String() != test.wantData {
+				t.Errorf("got data: %v, want: %v", buf.String(), test.wantData)
+			}
+			if s.offset != test.wantOffset {
```

Comment:
I have mixed feelings, too. The tests are too tightly coupled to implementation details. I see them as a tool for development which should be deleted as soon as they get in the way of a refactoring (which ideally would allow to write better tests).

With this kind of testing it is possible to easily simulate and test seek and write errors, and the tests run quickly as they do not have to touch the disk for each test case (cleanup of files would be required as well).

An integrated test using a reader cannot detect false assumptions about offsets and encoding, for example. Two errors might cancel out, but still result in an incorrect shp / dbf file.

Reply:
Yes, it is obvious that integration tests alone are not enough. Let's keep your test for now and add an integration test later. The interfaces you identified will make it simple to write this without requiring any disk accesses.